File: 696. Count Binary Substrings/_696.py
class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        pre = 0
        res = list()
        for i in range(1, len(s)):
            if s[i] != s[i-1]:
                res.append(i-pre)
                pre = i
        res.append(len(s)-pre)
        fin = 0
        for i in range(1,len(res)):
            fin += min(res[i], res[i-1])
        return fin

    def countBinarySubstrings1(self, s):
        s = [len(i) for i in s.replace('01', '0 1').replace('10', '1 0').split()]
        return sum(min(a, b) for a, b in zip(s, s[1:]))

a = Solution()
print(a.countBinarySubstrings("10101"))

File: 696. Count Binary Substrings/696.go
package leetcode

import(
	"math"
)

func countBinarySubstrings(s string) int {
	res := make([]int, 0, 1)
	pre := 0
	for i:=1; i<len(s); i++{
		if s[i] != s[i-1]{
			res = append(res, i-pre)
			pre = i
		}
	}
	res = append(res, len(s)-pre)

	pre = 0
	for i:=1; i<len(res);i++{
		if res[i]<res[i-1]{
			pre+=res[i]
		}else{
			pre+=res[i-1]
		}
	}
	return pre
}

func countBinarySubstrings1(s string) int {
	var cur, pre, res float64
    cur = 1  
	for i := 1; i < len(s); i++ {
		if s[i-1] == s[i] {
			cur++
		} else {
			res += math.Min(cur, pre)
			pre = cur
			cur = 1
		}
	}
	return int(res + math.Min(cur, pre))
}